precompiles/referenda/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+			track_id,
+			deciding_count
+		);
+
+		Ok(deciding_count.into())
+	}
+
+	#[precompile::public("trackIds()")]
+	#[precompile::view]
+	fn track_ids(handle: &mut impl PrecompileHandle) -> EvmResult<Vec<u16>> {
+		// Fetch data from runtime
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let track_ids: Vec<u16> = Runtime::Tracks::tracks()
+			.into_iter()
+			.filter_map(|x| {
+				if let Ok(track_id) = x.0.try_into() {
+					Some(track_id)
+				} else {
+					None
+				}
+			})
+			.collect();
+
+		Ok(track_ids)
+	}
+
+	#[precompile::public("trackInfo(uint16)")]
+	#[precompile::view]
+	fn track_info(
+		handle: &mut impl PrecompileHandle,
+		track_id: u16,
+	) -> EvmResult<(
+		UnboundedBytes,
+		U256,
+		U256,
+		U256,
+		U256,
+		U256,
+		U256,
+		UnboundedBytes,
+		UnboundedBytes,
+	)> {
+		// Fetch data from runtime
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+		let track_id: TrackIdOf<Runtime> = track_id
+			.try_into()
+			.map_err(|_| RevertReason::value_is_too_large("Track id type").into())
+			.in_field("trackId")?;
+		let tracks = Runtime::Tracks::tracks();
+		let index = tracks
+			.binary_search_by_key(&track_id, |x| x.0)
+			.unwrap_or_else(|x| x);
+		let track_info = &tracks[index].1;
+
+		Ok((
+			track_info.name.as_bytes().into(),
+			track_info.max_deciding.into(),
+			track_info.decision_deposit.into(),
+			track_info.prepare_period.into(),
+			track_info.decision_period.into(),
+			track_info.confirm_period.into(),
+			track_info.min_enactment_period.into(),
+			track_info.min_approval.encode().into(),
+			track_info.min_support.encode().into(),
+		))
+	}
+
+	/// Propose a referendum on a privileged action.
+	///
+	/// Parameters:
+	/// * track_id: The trackId for the origin from which the proposal is to be dispatched.
+	/// * proposal: The proposed runtime call.
+	/// * block_number: Block number at which proposal is dispatched.
+	#[precompile::public("submitAt(uint16,bytes,uint32)")]
+	fn submit_at(
+		handle: &mut impl PrecompileHandle,
+		track_id: u16,
+		proposal: BoundedBytes<GetCallDataLimit>,
+		block_number: u32,
+	) -> EvmResult {
+		let origin: GovOrigin = track_id.try_into().map_err(|_| {
+			RevertReason::custom("Origin does not exist for TrackId").in_field("trackId")
+		})?;
+		let proposal_origin: Box<OriginOf<Runtime>> = Box::new(origin.into());
+		let proposal: BoundedCallOf<Runtime> = Bounded::Inline(
+			frame_support::BoundedVec::try_from(proposal.as_bytes().to_vec()).map_err(|_| {
+				RevertReason::custom("Proposal input is not a runtime call").in_field("proposal")
+			})?,
+		);
+		let enactment_moment = DispatchTime::At(block_number.into());
+
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+
+		let call = ReferendaCall::<Runtime>::submit {
+			proposal_origin,
+			proposal,
+			enactment_moment,
+		}
+		.into();
+
+		<RuntimeHelper<Runtime>>::try_dispatch(handle, Some(origin).into(), call)?;
```

Comment:
Is this the correct `origin` in this context?

Reply:
(this pattern is repeated below, so check elsewhere if needed)

Reply:
Oh, I see that `origin` is shadowed. That's kind of a scary thing to shadow here...

Reply:
followed suggestion